### PR TITLE
alternative workcenters

### PR DESCRIPTION
**Pull request by Whenrow**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/models/mrp_production.py [hunk 1]
```diff
@@ -610,54 +610,87 @@ def button_plan(self):
...
+
+            # extract already planned workorders
+            wos = self.env['mrp.workorder'].search([
+                ('workcenter_id', 'in', workcenters.ids),
+                ('state', 'in', ('ready', 'pending', 'progress')),
+                ('date_planned_finished', '>=', start_date.strftime(DEFAULT_SERVER_DATETIME_FORMAT))],
+                order="date_planned_finished"
+            )
+            best_start_date = start_date
+            best_finished_date = start_date + relativedelta(year=MAXYEAR)
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: datetime.max

File: addons/mrp/models/mrp_production.py [hunk 2]
```diff
@@ -610,54 +610,87 @@ def button_plan(self):
...
+                has_schedule = workcenter.resource_calendar_id.attendance_ids
+                if has_schedule:
+                    # planning 0 hours gives the start of the next attendance
+                    from_date = workcenter.resource_calendar_id.plan_hours(0, from_date, compute_leaves=True, resource=workcenter.resource_id)
+                    to_date = workcenter.resource_calendar_id.plan_hours(duration_expected / 60.0, from_date, compute_leaves=True, resource=workcenter.resource_id)
                 else:
-                    to_date = from_date + relativedelta(minutes=workorder.duration_expected)
-                # Check interval
-                for wo in wos:
+                    to_date = from_date + relativedelta(minutes=duration_expected)
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: you could use dateutils

File: addons/mrp/models/mrp_production.py [hunk 3]
```diff
@@ -610,54 +610,87 @@ def button_plan(self):
...
+                # Check intervals
+                for wo in wos.filtered(lambda wo: wo.workcenter_id == workcenter):
+                    if best_finished_date < fields.Datetime.from_string(wo.date_planned_finished):
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: remove fields.Datetime.from_string(

File: addons/mrp/models/mrp_production.py [hunk 4]
```diff
@@ -610,54 +610,87 @@ def button_plan(self):
...
-                workcenter = workorder.workcenter_id
-                wos = WorkOrder.search([('workcenter_id', '=', workcenter.id), ('date_planned_finished', '<>', False),
-                                        ('state', 'in', ('ready', 'pending', 'progress')),
-                                        ('date_planned_finished', '>=', start_date.strftime(DEFAULT_SERVER_DATETIME_FORMAT))], order='date_planned_start')
+        start_date = self._get_start_date()
+        for workorder in self.workorder_ids:
+            workcenters = workorder.workcenter_id | workorder.workcenter_id.alternative_workcenter_ids
+
+            # extract already planned workorders
+            wos = self.env['mrp.workorder'].search([
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: search_read or read_group for perf

File: addons/mrp/models/mrp_production.py [hunk 5]
```diff
@@ -610,54 +610,87 @@ def button_plan(self):
...
+            # Compute start_date for next workorder
+            if workorder.next_work_order_id:
                 if (workorder.operation_id.batch == 'no') or (workorder.operation_id.batch_size >= workorder.qty_production):
-                    start_date = to_date
+                    start_date = best_finished_date
                 else:
-                    qty = min(workorder.operation_id.batch_size, workorder.qty_production)
-                    cycle_number = math.ceil(qty / workorder.production_id.product_qty / workcenter.capacity)
+                    workcenter = workorder.workcenter_id
+                    qty = min(workorder.operation_id.batch_size, workorder.qty_producing)
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: qty = workorder.operation_id.batch_size

File: addons/mrp/models/mrp_production.py [hunk 6]
```diff
@@ -610,54 +610,87 @@ def button_plan(self):
...
+                        'workcenter_id': workcenter.id,
+                        'capacity': workcenter.capacity,
+                        'duration_expected': duration_expected,
+                    }
+
+            workorder.write(best_values)
+
+            # Compute start_date for next workorder
+            if workorder.next_work_order_id:
                 if (workorder.operation_id.batch == 'no') or (workorder.operation_id.batch_size >= workorder.qty_production):
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: qty_producing

File: addons/mrp/models/mrp_production.py [hunk 7]
```diff
@@ -610,54 +610,87 @@ def button_plan(self):
...
-                    cycle_number = math.ceil(qty / workorder.production_id.product_qty / workcenter.capacity)
+                    workcenter = workorder.workcenter_id
+                    qty = min(workorder.operation_id.batch_size, workorder.qty_producing)
+                    cycle_number = math.ceil(qty / workorder.qty_producing / workcenter.capacity)
                     duration = workcenter.time_start + cycle_number * workorder.operation_id.time_cycle * 100.0 / workcenter.time_efficiency
-                    to_date = workcenter.resource_calendar_id.attendance_ids and workcenter.resource_calendar_id.plan_hours(duration / 60.0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                    if not to_date:
-                        start_date = from_date + relativedelta(minutes=duration)
+                    if has_schedule:
+                        start_date = workcenter.resource_calendar_id.plan_hours(duration / 60.0, start_date, compute_leaves=True, resource=workcenter.resource_id)
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: from_date

File: addons/mrp/models/mrp_production.py [hunk 8]
```diff
@@ -610,54 +610,87 @@ def button_plan(self):
...
+                    else:
+                        start_date = start_date + relativedelta(minutes=duration)
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: best_start_date

File: addons/mrp/models/mrp_production.py [hunk 9]
```diff
@@ -610,54 +608,90 @@ def button_plan(self):
...
+
                 from_date = start_date
-                to_date = workcenter.resource_calendar_id.attendance_ids and workcenter.resource_calendar_id.plan_hours(workorder.duration_expected / 60.0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                if to_date:
-                    if not from_date_set:
-                        # planning 0 hours gives the start of the next attendance
-                        from_date = workcenter.resource_calendar_id.plan_hours(0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                        from_date_set = True
+                # Check if the workcenter is subject to a schedule
+                has_schedule = workcenter.resource_calendar_id.attendance_ids
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: You could create a variable for workcenter_calendar instead of this one, it would be more usefull

File: addons/mrp/models/mrp_production.py [hunk 10]
```diff
@@ -610,54 +608,90 @@ def button_plan(self):
...
-                        if not to_date:
-                            to_date = from_date + relativedelta(minutes=workorder.duration_expected)
-                workorder.write({'date_planned_start': from_date, 'date_planned_finished': to_date})
-
-                if (workorder.operation_id.batch == 'no') or (workorder.operation_id.batch_size >= workorder.qty_production):
-                    start_date = to_date
+                    to_date = date_utils.add(from_date, minutes=duration_expected)
+                # Check intervals
+                for date_start, date_finished in wos.get(workcenter.id, []):
+                    if best_finished_date < date_finished:
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]:
```
if best_finished_date < date_finished or to_date <= date_start
   break
else:
```

File: addons/mrp/models/mrp_production.py [hunk 11]
```diff
@@ -610,54 +608,90 @@ def button_plan(self):
...
+                        # workorders are ordered by date. There is no need to check the other ones
+                        break
+                    if to_date > date_start:
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: # The current time slot (from_date, to_date) impinge on an existing workorder

File: addons/mrp/models/mrp_production.py (hunk 10 above)
Comment (amoyaux) [reply addons/mrp/models/mrp_production.py]: Or you could so something like that
``` python3
if best_finished_date < date_finished or to_date > best_finished_date
   break
if to_date > date_start:
   next time slot
else:
   best_finished_date = to_date
   ....
```

File: addons/mrp/models/mrp_production.py [hunk 12]
```diff
@@ -687,35 +721,28 @@ def _workorders_create(self, bom, bom_data):
                     BoMs
         """
         workorders = self.env['mrp.workorder']
-        bom_qty = bom_data['qty']
 
         # Initial qty producing
+        quantity = max(self.product_qty - sum(self.move_finished_ids.filtered(lambda move: move.product_id == self.product_id).mapped('quantity_done')), 0)
+        if self.product_uom_id.uom_type != 'reference':
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: Maybe you should ask to lap for this... It seems weird

File: addons/mrp/models/mrp_production.py [hunk 13]
```diff
@@ -624,54 +624,94 @@ def button_plan(self):
...
+                ],
+                fields=['workcenter_id', 'date_planned_start', 'date_planned_finished'],
+                order="workcenter_id",
+            )
+            # group workorders (represented by dates) by workcenter
+            groupby_workcenter = groupby(wos1, key=lambda wo: wo['workcenter_id'][0])
+            wos = {}
+            for wc, dates in groupby_workcenter:
+                dates = [(date['date_planned_start'], date['date_planned_finished']) for date in dates]
+                wos[wc] = sorted(dates, key=itemgetter(1))
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: try in search_read

File: addons/mrp/models/mrp_production.py [hunk 14]
```diff
@@ -624,54 +624,94 @@ def button_plan(self):
...
-                workorder.write({'date_planned_start': from_date, 'date_planned_finished': to_date})
-
-                if (workorder.operation_id.batch == 'no') or (workorder.operation_id.batch_size >= workorder.qty_production):
-                    start_date = to_date
+                    to_date = date_utils.add(from_date, minutes=duration_expected)
+                # Check intervals
+                for date_start, date_finished in wos.get(workcenter.id, []):
+                    if best_finished_date < date_finished or to_date <= date_start:
+                        # 1.workorders are ordered by date. There is no need to check the other ones
+                        # or 2.The current time slot (from_date, to_date) intersect an existing workorder slot
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py]: remove or

File: addons/mrp/models/mrp_production.py [hunk 15]
```diff
@@ -624,54 +623,72 @@ def button_plan(self):
...
-                    if not from_date_set:
-                        # planning 0 hours gives the start of the next attendance
-                        from_date = workcenter.resource_calendar_id.plan_hours(0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                        from_date_set = True
+        start_date = self._get_start_date()
+        for workorder in self.workorder_ids:
+            workcenters = workorder.workcenter_id | workorder.workcenter_id.alternative_workcenter_ids
+
+            best_finished_date = datetime.datetime.max
+            best_values = {}
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: vals = {}

File: addons/mrp/models/mrp_production.py [hunk 16]
```diff
@@ -624,54 +623,72 @@ def button_plan(self):
...
+            for workcenter in workcenters:
+                # compute theoritical duration
+                operation = workorder.operation_id
+                cycle_number = float_round(workorder.qty_producing / workcenter.capacity, precision_digits=0, rounding_method='UP')
+                duration_expected = workcenter.time_start + workcenter.time_stop + cycle_number * operation.time_cycle * 100.0 / workcenter.time_efficiency
+
+                # get first free slot
+                # planning 0 hours gives the start of the next attendance
+                from_date = workcenter.resource_calendar_id.plan_hours(0, start_date, compute_leaves=True, resource=workcenter.resource_id, domain=[('time_type', 'in', ['leave', 'other'])])
+                if from_date:
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]:
``` python
# If the workcenter is unavailable, try on the next one
if from_date is False:
   continue
```

File: addons/mrp/models/mrp_production.py [hunk 17]
```diff
@@ -624,54 +623,72 @@ def button_plan(self):
...
+                    best_start_date = from_date
+                    best_finished_date = to_date
+                    best_workcenter = workcenter
+                    best_values = {
+                        'workcenter_id': workcenter.id,
+                        'capacity': workcenter.capacity,
+                        'duration_expected': duration_expected,
+                    }
+
+            if best_finished_date == datetime.datetime.max:
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: add a comment
# If all workcenters are unavailable, raise.

File: addons/mrp/models/mrp_production.py [hunk 18]
```diff
@@ -624,54 +623,72 @@ def button_plan(self):
...
-                                        ('date_planned_finished', '>=', start_date.strftime(DEFAULT_SERVER_DATETIME_FORMAT))], order='date_planned_start')
-                from_date = start_date
-                to_date = workcenter.resource_calendar_id.attendance_ids and workcenter.resource_calendar_id.plan_hours(workorder.duration_expected / 60.0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                if to_date:
-                    if not from_date_set:
-                        # planning 0 hours gives the start of the next attendance
-                        from_date = workcenter.resource_calendar_id.plan_hours(0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                        from_date_set = True
+        start_date = self._get_start_date()
+        for workorder in self.workorder_ids:
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: you shouldn't rely on this! self.workorder_ids may not return the workorders correctly ordered
either you sort them by next_workorder_id before looping over or you enforce an order on the mrp.workorder model (currently none is enforced)

File: addons/mrp/models/mrp_production.py [hunk 19]
```diff
@@ -624,54 +623,72 @@ def button_plan(self):
...
+                        'capacity': workcenter.capacity,
+                        'duration_expected': duration_expected,
+                    }
+
+            if best_finished_date == datetime.datetime.max:
+                raise UserError(_('Impossible to plan the workorder. Please check the workcenter availabilities'))
+
+            # Compute start_date for next workorder if operation allow batch producing
+            if workorder.next_work_order_id:
+                if (workorder.operation_id.batch == 'no') or (workorder.operation_id.batch_size >= workorder.qty_producing):
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: useless parens

File: addons/mrp/models/mrp_production.py [hunk 20]
```diff
@@ -624,54 +623,72 @@ def button_plan(self):
...
+            # Create leave on choosen workcenter calendar
+            leave = self.env['resource.calendar.leaves'].create({
+                'name': self.name + ' - ' + workorder.name,
+                'calendar_id': best_workcenter.resource_calendar_id.id,
+                'date_from': best_start_date,
+                'date_to': best_finished_date,
+                'resource_id': best_workcenter.resource_id.id,
+                'time_type': 'other'
+            })
+            best_values.update({'leave_id': leave.id})
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: vals['leave_id'] = leave.id

File: addons/mrp/models/mrp_production.py [hunk 21]
```diff
@@ -624,54 +623,72 @@ def button_plan(self):
...
+                    best_values = {
+                        'workcenter_id': workcenter.id,
+                        'capacity': workcenter.capacity,
+                        'duration_expected': duration_expected,
+                    }
+
+            if best_finished_date == datetime.datetime.max:
+                raise UserError(_('Impossible to plan the workorder. Please check the workcenter availabilities'))
+
+            # Compute start_date for next workorder if operation allow batch producing
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: add a comment
instantiate start_date for the next iteration

File: addons/mrp/models/mrp_workorder.py [hunk 22]
```diff
@@ -114,6 +123,26 @@ class MrpWorkorder(models.Model):
         'finished_workorder_id', string='By-products')
     allowed_lots_domain = fields.One2many(comodel_name='stock.production.lot', compute="_compute_allowed_lots_domain")
 
+    @api.depends('leave_id')
+    def _compute_dates_planned(self):
+        for wo in self:
+            wo.update({
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: why update in a compute?

File: addons/mrp/models/mrp_workorder.py [hunk 23]
```diff
@@ -114,6 +123,26 @@ class MrpWorkorder(models.Model):
         'finished_workorder_id', string='By-products')
     allowed_lots_domain = fields.One2many(comodel_name='stock.production.lot', compute="_compute_allowed_lots_domain")
 
+    @api.depends('leave_id')
+    def _compute_dates_planned(self):
+        for wo in self:
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: for workorder in self

File: addons/mrp/models/mrp_workorder.py [hunk 24]
```diff
@@ -114,6 +123,26 @@ class MrpWorkorder(models.Model):
         'finished_workorder_id', string='By-products')
     allowed_lots_domain = fields.One2many(comodel_name='stock.production.lot', compute="_compute_allowed_lots_domain")
 
+    @api.depends('leave_id')
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]: add a comment: 
```
Both `date_planned_start` and `date_planned_finished` are related fields on `leave_id`. Let's say
we slide a workorder on a gantt view, a single call to write is made with both fields. As the ORM
doesn't batch the write on related fields and instead makes multiple call, a constraint is raised. That's
why the four next methods are needed.
```

File: addons/resource/security/resource_security.xml [hunk 25]
```diff
@@ -16,7 +16,7 @@
         <field name="name">resource.calendar.leaves: employee modifies own</field>
         <field name="model_id" ref="model_resource_calendar_leaves"/>
         <field name="groups" eval="[(4, ref('base.group_user'))]"/>
-        <field name="domain_force">[('resource_id.user_id', '=', user.id)]</field>
+        <field name="domain_force">[('resource_id.user_id', 'in', [False, user.id])]</field>
```
Comment (sle-odoo) [inline addons/resource/security/resource_security.xml]: split this part in another commit
[ref] resource: relax ir rule on resource.calendar.leaves

File: addons/mrp/models/mrp_production.py (hunk 18 above)
Comment (Whenrow) [reply addons/mrp/models/mrp_production.py]: Seems that the default order is good enough: 
https://github.com/odoo/odoo/blob/086b9ad879664448d8a4c9102b2cf158def04444/odoo/models.py#L249

File: addons/mrp/models/mrp_workorder.py [hunk 26]
```diff
@@ -114,6 +123,33 @@ class MrpWorkorder(models.Model):
...
+    # That's why the compute and set methods are needed. to ensure the dates are updated
+    # in the same time. The two next search method are needed as the field are non stored and
+    # not direct related fields.
+    @api.depends('leave_id')
+    def _compute_dates_planned(self):
+        for workorder in self:
+            workorder.update({
+                'date_planned_start': workorder.leave_id.date_from,
+                'date_planned_finished': workorder.leave_id.date_to
+            })
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]:
``` python
for workorder in self:
    workorder.date_planned_start = workorder.leave_id.date_from
    workorder. date_planned_finished = workorder.leave_id

File: addons/mrp/models/mrp_workorder.py [hunk 27]
```diff
@@ -114,6 +123,33 @@ class MrpWorkorder(models.Model):
...
+
+    def _set_dates_planned(self):
+        self.leave_id.write({
+            'date_from': self.date_planned_start,
+            'date_to': self.date_planned_finished,
+        })
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_workorder.py]:
``` python
date_from = self.date_planned_start
date_to = self.date_planned_finished
for workorder in self:
    workorder.write({
        'date_from': date_from,
        'date_to': date_to,
    })
```

File: addons/mrp/models/mrp_production.py [hunk 28]
```diff
@@ -624,54 +623,74 @@ def button_plan(self):
...
-                        from_date = workcenter.resource_calendar_id.plan_hours(0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                        from_date_set = True
+        start_date = self._get_start_date()
+        for workorder in self.workorder_ids:
+            workcenters = workorder.workcenter_id | workorder.workcenter_id.alternative_workcenter_ids
+
+            best_finished_date = datetime.datetime.max
+            vals = {}
+            for workcenter in workcenters:
+                # compute theoritical duration
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: theoretical

File: addons/mrp/models/mrp_production.py [hunk 29]
```diff
@@ -624,54 +623,74 @@ def button_plan(self):
...
+                operation = workorder.operation_id
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: time_cycle = workorder.operation_id.time_cycle

File: addons/mrp/models/mrp_production.py [hunk 30]
```diff
@@ -624,54 +623,74 @@ def button_plan(self):
...
+                cycle_number = float_round(workorder.qty_producing / workcenter.capacity, precision_digits=0, rounding_method='UP')
+                duration_expected = workcenter.time_start + workcenter.time_stop + cycle_number * operation.time_cycle * 100.0 / workcenter.time_efficiency
+
+                # get first free slot
+                # planning 0 hours gives the start of the next attendance
+                from_date = workcenter.resource_calendar_id.plan_hours(0, start_date, compute_leaves=True, resource=workcenter.resource_id, domain=[('time_type', 'in', ['leave', 'other'])])
+                # If the workcenter is unavailable, try planning on the next one
+                if from_date is False:
+                    continue
                 else:
```
Comment (sle-odoo) [inline addons/mrp/models/mrp_production.py]: remove, unindent next line

File: addons/mrp/tests/test_workorder_operation.py [hunk 31]
```diff
@@ -2,18 +2,55 @@
...
+            'product_id': self.product_4.id,
+            'product_tmpl_id': self.product_4.product_tmpl_id.id,
+            'product_uom_id': self.uom_unit.id,
+            'product_qty': 4.0,
+            'routing_id': self.routing_1.id,
+            'type': 'normal',
+            'bom_line_ids': [
+                (0, 0, {'product_id': self.product_2.id, 'product_qty': 2}),
+                (0, 0, {'product_id': self.product_1.id, 'product_qty': 4})
+            ]})
```
Comment (sle-odoo) [inline addons/mrp/tests/test_workorder_operation.py]: you just added ~10 seconds on runbot, we could maybe avoid that by moving all these master data creation in the setupclass? can you test?

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff
